File: genql/compilers/python_compiler.py
"""
genql.compilers.python_compiler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiles COMPUTE intents into Python function definitions.

Supported body keys
-------------------
function : str
    Name for the generated function.
params : list[str]
    Ordered parameter names.
expression : str
    A single Python expression that forms the function body.
    Use this for pure transformations (the common case).
steps : list[str]
    Ordered list of statement lines.  Use instead of ``expression``
    when multiple statements are required.
returns : str  (optional)
    Type annotation for the return value.
"""

from __future__ import annotations

from typing import List

from genql.compilers.base import Compiler
from genql.intent import Intent, IntentKind


class PythonCompiler(Compiler):
    """Translates :attr:`IntentKind.COMPUTE` intents to Python source."""

    @property
    def target_language(self) -> str:
        return "python"

    @property
    def supported_kinds(self) -> List[IntentKind]:
        return [IntentKind.COMPUTE]

    def compile(self, intent: Intent) -> str:
        body = intent.body
        fn_name = str(body.get("function", intent.name))
        params: List[str] = list(body.get("params") or [])
        returns: str = body.get("returns", "")
        expression: str = body.get("expression", "")
        steps: List[str] = list(body.get("steps") or [])

        # Build signature
        sig = ", ".join(params)
        ret_ann = f" -> {returns}" if returns else ""
        header = f"def {fn_name}({sig}){ret_ann}:"

        # Build body lines
        if steps:
            body_lines = steps
        elif expression:
            body_lines = [f"return {expression}"]
        else:
            body_lines = ["pass"]

        indented = ["    " + line for line in body_lines]
        return "\n".join([header] + indented)

File: genql_ruby/lib/genql/compilers/python_compiler.rb
# genql/compilers/python_compiler
#
# Compiles COMPUTE intents into Python function definitions.
#
# Supported body keys
# -------------------
# function   : String       — name for the generated function
# params     : Array<String> — ordered parameter names
# expression : String       — a single Python expression (pure functions)
# steps      : Array<String> — ordered statement lines (multi-statement bodies)
# returns    : String       — type annotation for the return value (optional)

require_relative "base"
require_relative "../intent"

module Genql
  module Compilers
    class PythonCompiler < Compiler
      def target_language
        "python"
      end

      def supported_kinds
        [IntentKind::COMPUTE]
      end

      def compile(intent)
        body       = intent.body
        fn_name    = body_get(body, "function") || intent.name
        params     = Array(body_get(body, "params"))
        returns    = body_get(body, "returns") || ""
        expression = body_get(body, "expression") || ""
        steps      = Array(body_get(body, "steps"))

        sig      = params.join(", ")
        ret_ann  = returns.empty? ? "" : " -> #{returns}"
        header   = "def #{fn_name}(#{sig})#{ret_ann}:"

        body_lines =
          if steps.any?
            steps
          elsif !expression.empty?
            ["return #{expression}"]
          else
            ["pass"]
          end

        indented = body_lines.map { |line| "    #{line}" }
        ([header] + indented).join("\n")
      end

      private

      def body_get(body, key)
        body[key.to_sym] || body[key.to_s]
      end
    end
  end
end
